### PR TITLE
Add ability to weight training samples and predefine training/validation splits

### PR DESCRIPTION
**Pull request by ledbetdr**
Enables the ability to utilize sklearn's PredefinedSplit to allow specific training/validation splits by hijacking the eval_size parameter. This could also be fairly easily split into its own parameter.

More involved are the changes to allow weights to be specified during the fit routine to leverage the refactored aggregate functionality in Lasagne. Removes use of the now deprecated Objective object.

### CODE REVIEW

**Comment by dnouri** [comment]
Thanks for your work.  I would have an easier time to digest this pull request if it were split into two parts: one for the aggregate/weights/objective, another for PredefinedSplit.

Until now, we've recommended people to subclass and overwrite `train_test_split` if they need another train/validation split.  It was suggested elsewhere that a function with the same signature as `train_test_split` should be passed as a parameter to `NeuralNet`, instead of forcing users to subclass.  If you think `PredefinedSplit` would be more convenient than that, then please explain.

As for the removal of Objective: it's long overdue.  But I think we want a convenient replacement that allows losses to be calculated using outputs of potentially multiple layers.  See here: https://github.com/dnouri/nolearn/issues/105#issuecomment-106774786

As you saw, there's also test failures caused by your changes.  Although these have lower priority than figuring out the questions from above, to get your pull request merged, you'll need to make sure that all tests run, and that you add tests for any additional functionality.  (Click on Details in the line that says "Some checks have failed" in red.)

**Comment by ledbetdr** [comment]
The build checks that fail are caused by the inability to import aggregate from lasagne. I assume it's using an out-of-date version of lasagne, but I'm not sure how to change that? 

I'm definitely in favor of a re-factored train_test_split. Could we simply pass an sklearn KFold object to it so that users don't have to worry about subclassing or creating an entire new function? I think that should be fairly straight forward and would enable all the flexibility of the KFold objects (such as simple additional kargs).

I consider #105 a separate issue. Questions such as - can each objective function have differing weights per event? differing overall weights? It's not as straight forward of a change. The change here would become a base for such a change which demonstrates how a single weight array can be carried through and aggregated for a single loss function.

**Comment by cancan101** [comment]
I think this PR could great benefit from #118.

**Comment by dnouri** [comment]
On Mi, Jul 15, 2015 at 11:59 , ledbetdr wrote:
> The build checks that fail are caused by the inability to import 
> aggregate from lasagne. I assume it's using an out-of-date version of 
> lasagne, but I'm not sure how to change that?
> 
See here for how to upgrade the Lasagne version: 
https://github.com/dnouri/nolearn/blob/feature-remove-objective-class/requirements.txt#L6

(You may want to base your work on that branch.)
> I'm definitely in favor of a re-factored train_test_split. Could we 
> simply pass an sklearn KFold object to it so that users don't have to 
> worry about subclassing or creating an entire new function? I think 
> that should be fairly straight forward and would enable all the 
> flexibility of the KFold objects (such as simple additional kargs).
> 
Using the scikit-learn `train_test_split` interface with regard to 
return values might be easier to grasp and implement.  KFold and 
friends aren't a great match here, because they all assume you want a 
number of folds (for cross-validation).  It's why we have to use this 
funny `next(iter(kfold))`expression when we use them in method 
`train_test_split` right now.

I think that moving `train_test_split` out into a parameter is a good 
idea, and #118, which attempts to implement that is almost there.

**Comment by dnouri** [comment]
I put together an alternative implementation for factoring out `train_test_split` here: #125.

**Comment by dnouri** [comment]
So the first part of this pull request became obsolete with #125, which is good news.

If you're still up for it, please open a new pull request or update this one to only contain the changes needed for the sample weights feature.  (And note that the `objective` interface has changed as well in #124.)

**Comment by ledbetdr** [comment]
I'll try to find time later this week to rebase the changes on top of 125.

**Comment by dnouri** [comment]
Closing due to inactivity.
